### DISCUSSION TITLE
`run.sh` not found when running `docker-compose up`

### DISCUSSION THREAD

**Discussion by DannyVarod** [gnzsnz/ib-gateway-docker#192] [Q&A] [answered]
Hi, when running `docker-compose up` with the env vars configured to `ib-gateway` in `paper` mode, I get the following error message:

```
algo-trader-ib-gateway-1  | exec /home/ibgateway/scripts/run.sh: no such file or directory
```

In the `Dockerfile` I see that the scripts are copied with the following command:

```
COPY ./scripts /root/scripts
```

So I'm not sure how the `run.sh` script is supposed to reach the `/home/ibgateway/scripts` folder.

Thanks.

**Comment by gnzsnz** [comment]
Hi,

it's not easy to answer with the information available. but i will try to give you some pointers.

If I pull the images

```bash
docker pull ghcr.io/gnzsnz/ib-gateway:stable
docker pull ghcr.io/gnzsnz/ib-gateway:latest
```

and then I run

```bash
docker run --rm -it ghcr.io/gnzsnz/ib-gateway:stable ls -l /home/ibgateway/scripts
total 24
-rwxr-xr-x 1 ibgateway ibgateway 6700 Oct 25 09:05 common.sh
-rwxr-xr-x 1 ibgateway ibgateway 4099 Oct 25 09:05 run.sh
-rwxr-xr-x 1 ibgateway ibgateway  371 Oct 25 09:05 run_socat.sh
-rwxr-xr-x 1 ibgateway ibgateway  410 Oct 25 09:05 run_ssh.sh

docker run --rm -it ghcr.io/gnzsnz/ib-gateway:latest ls -l /home/ibgateway/scripts
total 24
-rwxr-xr-x 1 ibgateway ibgateway 6700 Oct 22 11:06 common.sh
-rwxr-xr-x 1 ibgateway ibgateway 4099 Oct 22 11:06 run.sh
-rwxr-xr-x 1 ibgateway ibgateway  371 Oct 22 11:06 run_socat.sh
-rwxr-xr-x 1 ibgateway ibgateway  410 Oct 22 11:06 run_ssh.sh
```

The scripts are on the right place and the image should work without problem. As it does when i run it locally.

If you are building your image locally, you might have different results. But that will depend on your changes and I can't help there.

To answer your question regarding the scripts, they arrive to `/home/ibgateway/scripts` in [line 69](https://github.com/gnzsnz/ib-gateway-docker/blob/d438f931018d671d95f1d4efacd1b6a45e64b6c2/Dockerfile.template#L69).

Hope it helps.

**Reply by DannyVarod** [reply]
Hi, thanks for your fast reply, especially on a weekend.

The only edit I did after cloning the repo was to the environment variables in the `docker-compose.yml` file, after which I ran `docker-compose up` from the repo root, this file pulls the `ghcr.io/gnzsnz/ib-gateway:10.30` image, not a local build.

I read the readme, however, I suspect I'm missing some basic step that needs to be done before running docker-compose.
Was I supposed to build myself and change the `docker-compose.yml` to use the locally built version?

Following the example in your reply, I tried running `docker run --rm -it ghcr.io/gnzsnz/ib-gateway:10.30` to see if this works, and I got the following output:

```
exec /home/ibgateway/scripts/run.sh: no such file or directory
```

which is strange, since `docker run --rm -it ghcr.io/gnzsnz/ib-gateway:10.30 ls -l /home/ibgateway/scripts` outputs:

```
total 24
-rwxr-xr-x 1 ibgateway ibgateway 6949 Oct 23 05:34 common.sh
-rwxr-xr-x 1 ibgateway ibgateway 4272 Oct 23 05:34 run.sh
-rwxr-xr-x 1 ibgateway ibgateway  385 Oct 23 05:34 run_socat.sh
-rwxr-xr-x 1 ibgateway ibgateway  425 Oct 23 05:34 run_ssh.sh
```

I copy pasted the paths into search to make sure there is no typo I'm missing, the paths are identical in command, in DockerFile, in file list.

I also checked all parent folders' permissions, all have `r-x` for all users, so this doesn't seem to be a user permission issue.

**Comment by gnzsnz** [comment]
I'm not sure how you ended up with `ghcr.io/gnzsnz/ib-gateway:10.30` you should have `stable`

this is what i get when i do

```bash
$ docker run --rm -it ghcr.io/gnzsnz/ib-gateway:10.30
Unable to find image 'ghcr.io/gnzsnz/ib-gateway:10.30' locally
10.30: Pulling from gnzsnz/ib-gateway
Digest: sha256:f183ce6fa48ed16ab635963760b8af0dd0e40f884eb1b41719ca3af8fbe735dc
Status: Downloaded newer image for ghcr.io/gnzsnz/ib-gateway:10.30
*************************************************************************
.> Starting IBC/IB gateway
*************************************************************************
.> Starting Xvfb server
.> SSH tunnel disabled
.> Usign default Java heap size 768m.
.> VNC server disabled
.> Invalid TRADING_MODE:
```
which makes sense because no env is set.

if you are not modifying the image, there is no need to clone the repo, you can just download `docker-compose.yml` and create an `.env` file, you can use the example on README and just add your user and pass.

i would suggest to start with a clean state

```bash
# delete local images
docker rmi ghcr.io/gnzsnz/ib-gateway:10.30 ghcr.io/gnzsnz/ib-gateway:stable
# get new image
docker compose pull
# run
docker compose up
```

**Reply by DannyVarod** [reply]
The readme says create a `.env` in the root folder, does this mean the root folder of the host (i.e. not in the image) or am I supposed to use the `volumes` in the `docker-compose.yml` file to map a file on my machine to the `/.env` file in the image?

My host is a Windows machine, so the paths work differently, e.g. if I run git-bash, then the `/` folder would be where `git-bash.exe` is installed.

I assumed that feeding the environment variables such as username and password to the `docker-compose.yml` is enough and the `.env` file is only if I create a custom image and use the `DockerFile` file to copy the file into it.

**Reply by gnzsnz** [reply]
i suggest to check the docs https://docs.docker.com/compose/how-tos/environment-variables/variable-interpolation/, i can't explain it better than docker itself

**Reply by sanurielf** [reply]
I did have a similar issue and found that it was due to the encoding of my scripts files. I solved running the tool dos2unix. Maybe give it a try.

**Reply by DannyVarod** [reply, marked as answer]
@gnzsnz, changing version in `docker-compose.yml` from `10.30` to `stable` fixed this issue.
Thanks.

@sanurielf, since I didn't build the docker image myself, the file encoding shouldn't matter, as the only file I was using was `docker-compose.yml`, I did try changing the encoding from UTF-8 to ASCII and EOF to LF for that file, but there was no effect. You can easily change these in VSC. Thanks.
